cmd/batch-handlers.go

```diff
@@ -0,0 +1,1068 @@
+	if ri == nil {
+		return errInvalidArgument
+	}
+	now := UTCNow()
+	if now.Sub(ri.LastUpdate) >= duration {
+		if serverDebugLog {
+			console.Debugf("batchReplicate: persisting batchReplication info on drive: threshold:%s, batchRepl:%#v\n", now.Sub(ri.LastUpdate), ri)
+		}
+		ri.LastUpdate = now
+		ri.Version = batchReplVersionV1
+		return ri.save(ctx, api, jobLocation)
+	}
+	return nil
+}
+
+func (ri *batchJobInfo) trackCurrentBucketObject(bucket string, info ObjectInfo, failed bool) {
+	if ri == nil {
+		return
+	}
+	ri.Bucket = bucket
+	ri.Object = info.Name
+	ri.countItem(info.Size, failed)
+}
+
+// Start start the batch replication job, resumes if there was a pending job via "job.ID"
+func (r *BatchJobReplicateV1) Start(ctx context.Context, api ObjectLayer, job BatchJobRequest) error {
+	ri := &batchJobInfo{
+		JobID:     job.ID,
+		JobType:   string(job.Type()),
+		StartTime: job.Started,
+	}
+	if err := ri.load(ctx, api, job); err != nil {
+		return err
+	}
+	globalBatchJobsMetrics.save(job.ID, ri.clone())
+	lastObject := ri.Object
+
+	delay := job.Replicate.Flags.Retry.Delay
+	if delay == 0 {
+		delay = batchReplJobDefaultRetryDelay
+	}
+	rnd := rand.New(rand.NewSource(time.Now().UnixNano()))
+
+	skip := func(info FileInfo) (ok bool) {
+		if r.Flags.Filter.OlderThan > 0 && time.Since(info.ModTime) < r.Flags.Filter.OlderThan {
+			// skip all objects that are newer than specified older duration
+			return false
+		}
+
+		if r.Flags.Filter.NewerThan > 0 && time.Since(info.ModTime) >= r.Flags.Filter.NewerThan {
+			// skip all objects that are older than specified newer duration
+			return false
+		}
+
+		if !r.Flags.Filter.CreatedAfter.IsZero() && r.Flags.Filter.CreatedAfter.Before(info.ModTime) {
+			// skip all objects that are created before the specified time.
+			return false
+		}
+
+		if !r.Flags.Filter.CreatedBefore.IsZero() && r.Flags.Filter.CreatedBefore.After(info.ModTime) {
+			// skip all objects that are created after the specified time.
+			return false
+		}
+
+		return true
+	}
+
+	u, err := url.Parse(r.Target.Endpoint)
+	if err != nil {
+		return err
+	}
+
+	cred := r.Target.Creds
+
+	c, err := miniogo.NewCore(u.Host, &miniogo.Options{
+		Creds:     credentials.NewStaticV4(cred.AccessKey, cred.SecretKey, cred.SessionToken),
+		Secure:    u.Scheme == "https",
+		Transport: getRemoteInstanceTransport,
+	})
+	if err != nil {
+		return err
+	}
+
+	retryAttempts := ri.RetryAttempts
+	for attempts := 1; attempts <= retryAttempts; attempts++ {
+		ctx, cancel := context.WithCancel(ctx)
+
+		results := make(chan ObjectInfo, 100)
+		if err := api.Walk(ctx, r.Source.Bucket, r.Source.Prefix, results, ObjectOptions{
+			WalkMarker: lastObject,
+			WalkFilter: skip,
+		}); err != nil {
+			cancel()
+			// Do not need to retry if we can't list objects on source.
+			return err
+		}
+
+		for result := range results {
+			if result.DeleteMarker {
+				// delete-markers will never be 'replicated'
+				continue
```

Comment:
if delete marker is not replicated, all the versions below it in the stack should not be replicated - seems it will be simpler to just replicate the delete marker too for consistency

Reply:
We can't replicate the delete marker since our plan is to support non-MinIO targets as well

Reply:
actually my intention is to skip objects that have only "delete marker" left, let me fix that.